### PR TITLE
Add einsum

### LINKED ISSUE #2340: Add `einsum` for `Qobj`.
### Problem Description

Having a version of `einsum` working on `Qobj`'s dimensions would be useful (qutip/qutip-qip#225).
A dense implementation should be easy enough to develop using `qutip.core.dimension.to_tensor_rep`.

### Proposed Solution

Use `to_tensor_rep`, `from_tensor_rep` with `np.einsum` to create qutip version of einsum.

### Alternate Solutions

If would be even better to develop it without conversion to numpy array to allow support for sparse, jax, cupy, etc. matrices.
However this would probably not be an easy task doable in the scope of `good first issue`.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by fmayo96**
**Description**
Added a einsum, a numpy.einsum (Evaluates the Einstein summation convention on the operands.) version for Qobj dimensions.

**Related issues or PRs**
fixes #2340

** Question: Not sure if I should update the documentation and how. Just in case I did not modify the docs yet.

### CODE REVIEW

File: qutip/core/dimensions.py [hunk 1]
```diff
@@ -352,6 +352,31 @@ def _frozen(*args, **kwargs):
...
+    Returns
+    -------
+    Qobj (numpy.complex128)
+        Result of einsum as Qobj (numpy.complex128 if result is scalar)
+    """
+    operands_array = [to_tensor_rep(op) for op in operands]
+    result = np.einsum(subscripts, *operands_array)
+    if result.shape == ():
+        return result
+    dims = [[d] for d in result.shape]
```
Comment (Ericgig) [inline qutip/core/dimensions.py:376]: Not sure if this work well with compound dimensions.
Could you add tests with multiple qubits systems.

**Review by Ericgig — COMMENTED** [review]
Thank you for tackling this issue.
Other than confirming the dims are good in multiple qubits systems, it looks good.

**Comment by hodgestar** [comment]
Thanks @fmayo96.

Question: Should the `einsum` function be in `dimensions.py`? And should it be internal or exposed via the public API? It looks generically useful but what specific use cases do we have in mind?

**Comment by BoxiLi** [comment]
The simplest use case is the circuit simulation, where e.g. single qubit unitaries are multiplied to multiqubit states. For Hamiltonian dynamics I cannot immediately think of an application cause usually you need to sum many terms. Unless we develop a troterrizatiom solver.

**Comment by fmayo96** [comment]
I have corrected the code for compound systems and added tests for two and three qubits. I am not sure about which is the best place for this function, I made it to solve the Issue #2340. If you consider it fits better in a different place I will be happy to modify it

File: qutip/core/dimensions.py (hunk 1 above)
Comment (Ericgig) [reply qutip/core/dimensions.py:376]: Sometime like 
`dims = [[d for d in result.shape[:result.ndim // 2], [d for d in result.shape[result.ndim // 2:]]`
would keep the compound dimensions.
We should raise warning or error if the output ndim is odd.

**Review by Ericgig — COMMENTED** [review]
Sorry, but the first attempt was better.
We want the `einsum` on the dimensions, not the shape. 
With a 2 qubits operator, there should be 4 indices, 6 indices for 3 qubits, etc.
The output for compound system should also be compound.
We want support for operations that could not be done with any mix of `matmul`, `trace`, etc.

Here are some example of use:

```
einsum("ikjl,jm->ikml", [tensor(sigmaz(), sigmaz()), sigmaz()]) == tensor(qeye(2), sigmaz())
einsum("ijkl->kjil", [tensor(sigmam(), sigmaz())]) == tensor(sigmap(), sigmaz())
```

There are no place for it, a new file (`qutip/core/einsum.py`) would be good, but it's fine to keep in `dimensions.py` for now. We will move it later.

**Comment by fmayo96** [comment]
Oh, I am sorry, I have misunderstood what you wanted. I think that now I have fixed it. Thank you very much for the feedback!

File: qutip/core/dimensions.py [hunk 2]
```diff
@@ -352,6 +352,32 @@ def _frozen(*args, **kwargs):
...
+    -------
+    Qobj (numpy.complex128)
+        Result of einsum as Qobj (numpy.complex128 if result is scalar)
+    """
+    operands_array = [to_tensor_rep(op) for op in operands]
+    result = np.einsum(subscripts, *operands_array)
+    if result.shape == ():
+        return result
+    dims = [d for d in result.shape[:result.ndim // 2]],
+    [[d for d in result.shape[result.ndim // 2:]]]
```
Comment (Ericgig) [inline qutip/core/dimensions.py:377]:
```suggestion
    dims = [
        [d for d in result.shape[:result.ndim // 2]],
        [d for d in result.shape[result.ndim // 2:]]
    ]
```

**Review by Ericgig — APPROVED** [review]
Thank you

**Pull request merged**